<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.11">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Oren Bochman">
<meta name="dcterms.date" content="2025-07-02">
<meta name="keywords" content="Gaussian Processes, Kriging, Optimal Interpolation">
<meta name="description" content="Gaussian Processes Tutorial">

<title>Gaussian Processes for Regression</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="C6-L01_files/libs/clipboard/clipboard.min.js"></script>
<script src="C6-L01_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="C6-L01_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="C6-L01_files/libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="C6-L01_files/libs/quarto-html/popper.min.js"></script>
<script src="C6-L01_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="C6-L01_files/libs/quarto-html/anchor.min.js"></script>
<link href="C6-L01_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="C6-L01_files/libs/quarto-html/quarto-syntax-highlighting-09b140d2d032adf2aedb8b099be3ee13.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="C6-L01_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="C6-L01_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="C6-L01_files/libs/bootstrap/bootstrap-068d5c994a6974d0eba00eca07de929a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Gaussian Processes for Regression</h1>
<p class="subtitle lead">Bayesian Statistics - Nonparametric Methods</p>
  <div class="quarto-categories">
    <div class="quarto-category">Bayesian Statistics</div>
    <div class="quarto-category">Nonparametric Methods</div>
  </div>
  </div>

<div>
  <div class="description">
    Gaussian Processes Tutorial
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Oren Bochman </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 2, 2025</p>
    </div>
  </div>
  
    
  </div>
  

<div>
  <div class="keywords">
    <div class="block-title">Keywords</div>
    <p>Gaussian Processes, Kriging, Optimal Interpolation</p>
  </div>
</div>

</header>


<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>Credit
</div>
</div>
<div class="callout-body-container callout-body">
<p>the following material is based on the <a href="https://tamarabroderick.com/tutorial_2024_cps-fr.html">Gaussian Processes for Regression</a> and tutorial and code by Tamara Broderick. <a href="https://tamarabroderick.com/tutorial_2025_asa_cirs_webinar.html"></a> Note that the <a href="https://github.com/tbroderick/gps_tutorial/blob/2025asa_cirs/demos.ipynb">code</a> is under the <a href="">MIT license</a></p>
</div>
</div>
<section id="why-gps" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="why-gps">Why GPs</h2>

<div class="no-row-height column-margin column-container"><div id="fig-nonparametrics-gps1" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametrics-gps1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="C6-SL01.png" class="img-fluid figure-img" style="width:53mm" data-group="slides"></p>
<figcaption>why GP1</figcaption>
</figure>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-nonparametrics-gps1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: ocean current vector field [Ryan, Özgökmen 2023; Zewe 2023; Gonçalves et al 2019; Lodise et al 2020; Berlinghieri et al 2023]
</figcaption>
</figure>
</div><div id="fig-nonparametrics-gps2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametrics-gps2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="C6-SL02.png" class="img-fluid figure-img" style="width:53mm" data-group="slides"></p>
<figcaption>why GP2</figcaption>
</figure>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-nonparametrics-gps2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: rocket booster surrogate model [Gramacy,Lee 2008] [Gramacy 2020]
</figcaption>
</figure>
</div></div>
<ul>
<li>Why Gaussian processes (GPs)?</li>
<li>Example 1:
<ul>
<li>The ocean current (velocity vector field) varies by space &amp; time</li>
<li>Scientists get sparse observations of the current from buoys</li>
<li>Goal: estimate the current</li>
<li>c.f. [Ryan, Özgökmen 2023; Zewe 2023; Gonçalves et al 2019; Lodise et al 2020; Berlinghieri et al 2023]</li>
</ul></li>
<li>Challenges:
<ul>
<li>Sparse &amp; expensive data, not on a grid</li>
<li>Current is highly nonlinear but smooth in space-time</li>
<li>We want uncertainty quantification</li>
<li>c.f.
<ul>
<li>[Gramacy,Lee 2008]</li>
<li>[Gramacy 2020]</li>
</ul></li>
</ul></li>
<li>Example 2: Surrogate model
<ul>
<li>The lift force of a rocket booster varies as a function of
<ul>
<li>speed at re-entry,</li>
<li>angle of attack, and</li>
<li>sideslip angle</li>
</ul></li>
<li>Scientists can run expensive simulations at chosen input settings</li>
<li>Goal: estimate how lift varies as a function of input settings</li>
</ul></li>
<li>Example 3: learn (&amp; optimize) performance in machine learning as a function of tuning parameters
<ul>
<li>c.f. [Snoek et al 2012; 2015; Garnett 2023]</li>
</ul></li>
<li>Motif:
<ul>
<li>Sparse, noisy, costly but but smooth data.</li>
<li>Output may have a nonlinear relationship to the inputs.</li>
<li>Want uncertainty quantification.</li>
<li>Low-dimensional inputs.</li>
</ul></li>
<li>Bonus benefits:
<ul>
<li>Ease of use (software, tuning)</li>
<li>Supports optimization of outcome</li>
<li>Predictions &amp; uncertainties over derivatives &amp; integrals</li>
<li>Module in more-complex methods</li>
</ul></li>
</ul>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Important</span>⚡ Overthinking parameters in Bayesian non-parametric models
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>Let’s talk about the elephant in the room non-parametrics do have parameters!</p></li>
<li><p>The classical form of a distribution based on a few fixed number parameters is replaced in these by a notion similar to a euclidean line which can be arbitrary extended. This is the nature of the <strong>non-parametric</strong> parameter vector <span class="math inline">\(\theta\)</span>.</p></li>
<li><p>To make the analogy more precise and what the above and many examples illustrate is that we often are interested in some behavior Y but as we get more and more data X, we will notice that there are new types of behavior that we did not observe earlier.</p></li>
</ul>
</div>
</div>
</section>
<section id="more-examples" class="level2">
<h2 class="anchored" data-anchor-id="more-examples">More Examples:</h2>
<ul>
<li>new species are being daily</li>
<li>new friend set appear as social networks expand</li>
<li>new entities appear in images as the data set expands</li>
<li>new wikipedia articles are discovered as we click on wikilinks</li>
<li>new alleles are identified as we sequence more genomes</li>
</ul>
</section>
<section id="roadmap" class="level2">
<h2 class="anchored" data-anchor-id="roadmap">Roadmap</h2>
<ul>
<li>A Bayesian approach</li>
<li>What is a Gaussian process?
<ul>
<li>Popular version using a squared exponential kernel</li>
</ul></li>
<li>Gaussian process inference.
<ul>
<li>Prediction &amp; uncertainty quantification.</li>
</ul></li>
<li>Goal:
<ul>
<li>Learn the mechanism behind standard GPs to identify benefits and pitfalls</li>
</ul></li>
</ul>
</section>
<section id="a-bayesian-approach" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="a-bayesian-approach">A Bayesian approach</h2>

<div class="no-row-height column-margin column-container"><div id="fig-nonparametics-posterior" class="quarto-float quarto-figure quarto-figure-center anchored" data-group="slides">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametics-posterior-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-SL04.png" class="img-fluid figure-img" style="width:53mm" data-group="slides">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-nonparametics-posterior-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: posterior
</figcaption>
</figure>
</div><div id="fig-nonparametics-joint" class="quarto-float quarto-figure quarto-figure-center anchored" data-group="slides">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametics-joint-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-SL05.png" class="img-fluid figure-img" style="width:53mm" data-group="slides">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-nonparametics-joint-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: joint
</figcaption>
</figure>
</div></div>
<p><span class="math display">\[
\underbrace{
  \mathbb{P}r(unknowns \mid data) }_{
    \substack{
      \text{Given the data we’ve seen} \\
      \text{what do we know about } \\
      \text{the underlying function?}
    }
  }
\propto
\underbrace{
  \mathbb{P}r(data \mid unknowns)
  \mathbb{P}r(unknowns)}_{  
    \substack{
      \text{A (statistical) model} \\
      \text{that can generate} \\
      \text{functions and data} \\
      \text{of interest}
    }
  }
\]</span></p>
</section>
<section id="sec-multivariate-gaussian-locations" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="sec-multivariate-gaussian-locations">Multivariate Gaussian using locations</h2>

<div class="no-row-height column-margin column-container"><div id="fig-nonparametrics-multivariate-gaussian-locations" class="quarto-float quarto-figure quarto-figure-center anchored" data-group="slides" width="53mm">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametrics-multivariate-gaussian-locations-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-SL06.png" id="fig-nonparametrics-multivariate-gaussian-locations" class="img-fluid figure-img" style="width:53mm" data-group="slides">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-nonparametrics-multivariate-gaussian-locations-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5
</figcaption>
</figure>
</div></div><ul>
<li><span class="math inline">\(M =2\)</span> bivariate Gaussian <span class="math inline">\([y^{(1)} y^{(2)}] \sim \mathcal{N}(\boldsymbol{\mu}, \boldsymbol{\Sigma})\)</span></li>
<li>with <span class="math inline">\(\mu = [0,0]^\top\)</span> and <span class="math inline">\(k = \sigma^2 \begin{bmatrix} 1 &amp; \rho \\ \rho &amp; 1\end{bmatrix}\)</span> where <span class="math inline">\(\rho\)</span> is the correlation between the two dimensions.</li>
<li>What if we let the correlation <span class="math inline">\(\rho\)</span> depend on the input <span class="math inline">\(x\)</span>?
<ul>
<li>Let <span class="math inline">\(\rho(x)= \rho(\mid x^{(1)} - x^{(2)} \mid)\)</span></li>
<li>Where the correlation goes to 1 as the x’s get close</li>
<li>And goes to 0 as the x’s get far</li>
</ul></li>
<li>Next: Use a Similar setup but an M-long Gaussian instead of just a bivariate
<ul>
<li>We have M locations (need not be in order): <span class="math inline">\(X^{(m)} \in \mathbb{R}^d\)</span></li>
<li>We generate <span class="math inline">\([y^{(1)}, \ldots,y^{(m)}]^\top \sim \mathcal{N}(\boldsymbol{\mu}, \boldsymbol{K})\)</span></li>
<li>with <span class="math inline">\(\mu=0_M\)</span> and <span class="math inline">\(K\)</span> such that <span class="math inline">\(\mathbf{K}_{ij} = \sigma^2 \rho\mid (x^{(i)}, x^{(j)}\mid)\)</span></li>
<li>Let’s try <span class="math inline">\(\rho(\Delta)=\exp(-\tfrac{1}{2} \Delta^2)\)</span>
<ul>
<li><span class="math inline">\(\rho(0)=1\)</span> ,</li>
<li><span class="math inline">\(\lim_{\Delta \to \infty} \rho(\Delta)=0\)</span> and</li>
<li><span class="math inline">\(\rho(\Delta)\)</span> is monotonically decreasing</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="demo-1-draws-from-a-multivariate-normal-distribution" class="level2">
<h2 class="anchored" data-anchor-id="demo-1-draws-from-a-multivariate-normal-distribution">Demo 1 : Draws from a multivariate normal distribution</h2>
<div id="lst-demo-01-setup" class="cell listing quarto-float quarto-figure quarto-figure-left anchored" data-execution_count="1">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst quarto-uncaptioned" id="lst-demo-01-setup-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;1
</figcaption>
<div aria-describedby="lst-demo-01-setup-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy <span class="im">as</span> sp</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>rng <span class="op">=</span> np.random.RandomState()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.gaussian_process <span class="im">import</span> GaussianProcessRegressor</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.gaussian_process.kernels <span class="im">import</span> RBF</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
<div id="cell-fig-demo-01-gp-draws-1" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="annotated-cell-2"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-2-1"><a href="#annotated-cell-2-1" aria-hidden="true" tabindex="-1"></a>signal_stddev <span class="op">=</span> <span class="fl">1.0</span></span>
<span id="annotated-cell-2-2"><a href="#annotated-cell-2-2" aria-hidden="true" tabindex="-1"></a>kern <span class="op">=</span> signal_stddev<span class="op">**</span><span class="dv">2</span><span class="op">*</span>RBF(</span>
<span id="annotated-cell-2-3"><a href="#annotated-cell-2-3" aria-hidden="true" tabindex="-1"></a>  length_scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="1" onclick="event.preventDefault();" href="">1</a><span id="annotated-cell-2-4" class="code-annotation-target"><a href="#annotated-cell-2-4" aria-hidden="true" tabindex="-1"></a>  length_scale_bounds<span class="op">=</span><span class="st">"fixed"</span>)</span>
<span id="annotated-cell-2-5"><a href="#annotated-cell-2-5" aria-hidden="true" tabindex="-1"></a>n_samples <span class="op">=</span> <span class="dv">3</span></span>
<span id="annotated-cell-2-6"><a href="#annotated-cell-2-6" aria-hidden="true" tabindex="-1"></a>gp <span class="op">=</span> GaussianProcessRegressor(kernel<span class="op">=</span>kern)</span>
<span id="annotated-cell-2-7"><a href="#annotated-cell-2-7" aria-hidden="true" tabindex="-1"></a>numx<span class="op">=</span><span class="dv">100</span></span>
<span id="annotated-cell-2-8"><a href="#annotated-cell-2-8" aria-hidden="true" tabindex="-1"></a>xstart<span class="op">=-</span><span class="fl">0.5</span></span>
<span id="annotated-cell-2-9"><a href="#annotated-cell-2-9" aria-hidden="true" tabindex="-1"></a>xstop<span class="op">=</span><span class="fl">3.5</span></span>
<span id="annotated-cell-2-10"><a href="#annotated-cell-2-10" aria-hidden="true" tabindex="-1"></a>x<span class="op">=</span>np.linspace(start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,num<span class="op">=</span>numx)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="2" onclick="event.preventDefault();" href="">2</a><span id="annotated-cell-2-11" class="code-annotation-target"><a href="#annotated-cell-2-11" aria-hidden="true" tabindex="-1"></a>twosd<span class="op">=</span><span class="dv">2</span><span class="op">*</span>signal_stddev<span class="op">*</span>np.ones(numx)</span>
<span id="annotated-cell-2-12"><a href="#annotated-cell-2-12" aria-hidden="true" tabindex="-1"></a>plt.fill_between(x,twosd,<span class="op">-</span>twosd,color<span class="op">=</span><span class="st">"gray"</span>,alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="annotated-cell-2-13"><a href="#annotated-cell-2-13" aria-hidden="true" tabindex="-1"></a>xdraw<span class="op">=</span>np.array([<span class="fl">0.0</span>,<span class="fl">0.05</span>]).reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>)<span class="op">;</span> </span>
<span id="annotated-cell-2-14"><a href="#annotated-cell-2-14" aria-hidden="true" tabindex="-1"></a>y_all_draws<span class="op">=</span>gp.sample_y(xdraw,random_state<span class="op">=</span>rng,</span>
<span id="annotated-cell-2-15"><a href="#annotated-cell-2-15" aria-hidden="true" tabindex="-1"></a>                        n_samples<span class="op">=</span>n_samples)<span class="op">;</span> </span>
<span id="annotated-cell-2-16"><a href="#annotated-cell-2-16" aria-hidden="true" tabindex="-1"></a>marker_collection<span class="op">=</span>[<span class="st">"o"</span>,<span class="st">"^"</span>,<span class="st">"x"</span>]</span>
<span id="annotated-cell-2-17"><a href="#annotated-cell-2-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ydraw <span class="kw">in</span> <span class="bu">enumerate</span>(y_all_draws.T): </span>
<span id="annotated-cell-2-18"><a href="#annotated-cell-2-18" aria-hidden="true" tabindex="-1"></a>  plt.scatter(</span>
<span id="annotated-cell-2-19"><a href="#annotated-cell-2-19" aria-hidden="true" tabindex="-1"></a>    xdraw,ydraw,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="3" onclick="event.preventDefault();" href="">3</a><span id="annotated-cell-2-20" class="code-annotation-target"><a href="#annotated-cell-2-20" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span>marker_collection[np.mod(i,n_samples)])</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-2" data-target-annotation="4" onclick="event.preventDefault();" href="">4</a><span id="annotated-cell-2-21" class="code-annotation-target"><a href="#annotated-cell-2-21" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])</span>
<span id="annotated-cell-2-22"><a href="#annotated-cell-2-22" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))</span>
<span id="annotated-cell-2-23"><a href="#annotated-cell-2-23" aria-hidden="true" tabindex="-1"></a>pltaxes <span class="op">=</span> plt.gca()</span>
<span id="annotated-cell-2-24"><a href="#annotated-cell-2-24" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"x"</span>)</span>
<span id="annotated-cell-2-25"><a href="#annotated-cell-2-25" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">"y"</span>)</span>
<span id="annotated-cell-2-26"><a href="#annotated-cell-2-26" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(<span class="st">"Multivariate normal draws; correlation decreases as x distance increases"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-2" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="4" data-code-annotation="1">set the kernel</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="11,12" data-code-annotation="2">coloring 2 std dev range</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="20" data-code-annotation="3">plotting the draws</span>
</dd>
<dt data-target-cell="annotated-cell-2" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-2" data-code-lines="21,22,23,24,25,26" data-code-annotation="4">make the plot</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-draws-1" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-draws-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-draws-1-output-1.png" width="619" height="449" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-draws-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Multivariate normal 2 draws
</figcaption>
</figure>
</div>
</div>
</div>
<div id="cell-fig-demo-01-gp-draws-2" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="annotated-cell-3"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-3-1"><a href="#annotated-cell-3-1" aria-hidden="true" tabindex="-1"></a>signal_stddev <span class="op">=</span> <span class="fl">1.0</span></span>
<span id="annotated-cell-3-2"><a href="#annotated-cell-3-2" aria-hidden="true" tabindex="-1"></a>kern <span class="op">=</span> signal_stddev<span class="op">**</span><span class="dv">2</span><span class="op">*</span>RBF(</span>
<span id="annotated-cell-3-3"><a href="#annotated-cell-3-3" aria-hidden="true" tabindex="-1"></a>  length_scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="1" onclick="event.preventDefault();" href="">1</a><span id="annotated-cell-3-4" class="code-annotation-target"><a href="#annotated-cell-3-4" aria-hidden="true" tabindex="-1"></a>length_scale_bounds<span class="op">=</span><span class="st">"fixed"</span>)</span>
<span id="annotated-cell-3-5"><a href="#annotated-cell-3-5" aria-hidden="true" tabindex="-1"></a>gp <span class="op">=</span> GaussianProcessRegressor(kernel<span class="op">=</span>kern)</span>
<span id="annotated-cell-3-6"><a href="#annotated-cell-3-6" aria-hidden="true" tabindex="-1"></a>numx<span class="op">=</span><span class="dv">100</span></span>
<span id="annotated-cell-3-7"><a href="#annotated-cell-3-7" aria-hidden="true" tabindex="-1"></a>xstart<span class="op">=-</span><span class="fl">0.5</span></span>
<span id="annotated-cell-3-8"><a href="#annotated-cell-3-8" aria-hidden="true" tabindex="-1"></a>xstop<span class="op">=</span><span class="fl">3.5</span></span>
<span id="annotated-cell-3-9"><a href="#annotated-cell-3-9" aria-hidden="true" tabindex="-1"></a>x<span class="op">=</span>np.linspace(</span>
<span id="annotated-cell-3-10"><a href="#annotated-cell-3-10" aria-hidden="true" tabindex="-1"></a>  start<span class="op">=</span>xstart,</span>
<span id="annotated-cell-3-11"><a href="#annotated-cell-3-11" aria-hidden="true" tabindex="-1"></a>  stop<span class="op">=</span>xstop,</span>
<span id="annotated-cell-3-12"><a href="#annotated-cell-3-12" aria-hidden="true" tabindex="-1"></a>  num<span class="op">=</span>numx)</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="2" onclick="event.preventDefault();" href="">2</a><span id="annotated-cell-3-13" class="code-annotation-target"><a href="#annotated-cell-3-13" aria-hidden="true" tabindex="-1"></a>twosd<span class="op">=</span><span class="dv">2</span><span class="op">*</span>signal_stddev<span class="op">*</span>np.ones(numx)</span>
<span id="annotated-cell-3-14"><a href="#annotated-cell-3-14" aria-hidden="true" tabindex="-1"></a>plt.fill_between(x,</span>
<span id="annotated-cell-3-15"><a href="#annotated-cell-3-15" aria-hidden="true" tabindex="-1"></a>  twosd,<span class="op">-</span>twosd,</span>
<span id="annotated-cell-3-16"><a href="#annotated-cell-3-16" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"gray"</span>,</span>
<span id="annotated-cell-3-17"><a href="#annotated-cell-3-17" aria-hidden="true" tabindex="-1"></a>  alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="annotated-cell-3-18"><a href="#annotated-cell-3-18" aria-hidden="true" tabindex="-1"></a>xdraw<span class="op">=</span>np.array([<span class="fl">0.0</span>,<span class="fl">0.05</span>,<span class="fl">3.0</span>]).reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="annotated-cell-3-19"><a href="#annotated-cell-3-19" aria-hidden="true" tabindex="-1"></a>y_all_draws<span class="op">=</span>gp.sample_y(</span>
<span id="annotated-cell-3-20"><a href="#annotated-cell-3-20" aria-hidden="true" tabindex="-1"></a>  xdraw,random_state<span class="op">=</span>rng,n_samples<span class="op">=</span><span class="dv">3</span>)</span>
<span id="annotated-cell-3-21"><a href="#annotated-cell-3-21" aria-hidden="true" tabindex="-1"></a>marker_collection<span class="op">=</span>[<span class="st">"o"</span>,<span class="st">"^"</span>,<span class="st">"x"</span>]</span>
<span id="annotated-cell-3-22"><a href="#annotated-cell-3-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ydraw <span class="kw">in</span> <span class="bu">enumerate</span>(y_all_draws.T): </span>
<span id="annotated-cell-3-23"><a href="#annotated-cell-3-23" aria-hidden="true" tabindex="-1"></a>  plt.scatter(</span>
<span id="annotated-cell-3-24"><a href="#annotated-cell-3-24" aria-hidden="true" tabindex="-1"></a>    xdraw,</span>
<span id="annotated-cell-3-25"><a href="#annotated-cell-3-25" aria-hidden="true" tabindex="-1"></a>    ydraw,</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="3" onclick="event.preventDefault();" href="">3</a><span id="annotated-cell-3-26" class="code-annotation-target"><a href="#annotated-cell-3-26" aria-hidden="true" tabindex="-1"></a>    marker<span class="op">=</span>marker_collection[np.mod(i,<span class="dv">3</span>)])</span>
<span id="annotated-cell-3-27"><a href="#annotated-cell-3-27" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])</span>
<span id="annotated-cell-3-28"><a href="#annotated-cell-3-28" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-3" data-target-annotation="4" onclick="event.preventDefault();" href="">4</a><span id="annotated-cell-3-29" class="code-annotation-target"><a href="#annotated-cell-3-29" aria-hidden="true" tabindex="-1"></a>pltaxes <span class="op">=</span> plt.gca()</span>
<span id="annotated-cell-3-30"><a href="#annotated-cell-3-30" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(<span class="st">"Multivariate normal draws; correlation decreases as x distance increases"</span>) <span class="co"># &lt;4&gt;plot</span></span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-3" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="4" data-code-annotation="1">set the kernel</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="2">2</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="13,17" data-code-annotation="2">coloring 2 std dev range</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="3">3</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="26" data-code-annotation="3">plotting the draws</span>
</dd>
<dt data-target-cell="annotated-cell-3" data-target-annotation="4">4</dt>
<dd>
<span data-code-cell="annotated-cell-3" data-code-lines="29" data-code-annotation="4">make the plot</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-draws-2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-draws-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-draws-2-output-1.png" width="603" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-draws-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Multivariate normal three points unevenly spaced
</figcaption>
</figure>
</div>
</div>
</div>
<div id="cell-fig-demo-01-gp-draws-3" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>signal_stddev <span class="op">=</span> <span class="fl">1.0</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>kern <span class="op">=</span> signal_stddev<span class="op">**</span><span class="dv">2</span><span class="op">*</span>RBF(length_scale<span class="op">=</span><span class="fl">1.0</span>,length_scale_bounds<span class="op">=</span><span class="st">"fixed"</span>) <span class="co"># set the kernel</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>gp <span class="op">=</span> GaussianProcessRegressor(kernel<span class="op">=</span>kern)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>numx<span class="op">=</span><span class="dv">200</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>xstart<span class="op">=-</span><span class="fl">0.5</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>xstop<span class="op">=</span><span class="fl">3.5</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>x<span class="op">=</span>np.linspace(start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,num<span class="op">=</span>numx)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>twosd<span class="op">=</span><span class="dv">2</span><span class="op">*</span>signal_stddev<span class="op">*</span>np.ones(numx) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>plt.fill_between(x,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  twosd,<span class="op">-</span>twosd,color<span class="op">=</span><span class="st">"gray"</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  alpha<span class="op">=</span><span class="fl">0.3</span>) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>xdraw<span class="op">=</span>np.linspace(</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  num<span class="op">=</span>numx).reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>y_all_draws<span class="op">=</span>gp.sample_y(</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  xdraw,random_state<span class="op">=</span>rng,</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  n_samples<span class="op">=</span><span class="dv">3</span>) </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>marker_collection<span class="op">=</span>[<span class="st">"o"</span>,<span class="st">"^"</span>,<span class="st">"x"</span>]</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ydraw <span class="kw">in</span> <span class="bu">enumerate</span>(y_all_draws.T): </span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  plt.scatter(xdraw,ydraw,marker<span class="op">=</span>marker_collection[np.mod(i,<span class="dv">3</span>)]) </span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>  plt.plot(xdraw,ydraw) <span class="co"># plotting the draws</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>pltaxes <span class="op">=</span> plt.gca() <span class="co"># make the plot</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(<span class="st">"Multivariate normal draws; correlation decreases as x distance increases"</span>) <span class="co"># make the plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-draws-3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-draws-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-draws-3-output-1.png" width="603" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-draws-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Multivariate normal 50 point evenly spaced
</figcaption>
</figure>
</div>
</div>
</div>
<p>we just drew random functions from a type of “Gaussian process”!</p>
</section>
<section id="sec-gaussian-processes" class="level2">
<h2 class="anchored" data-anchor-id="sec-gaussian-processes">Gaussian processes</h2>
<ul>
<li>Definition: “A Gaussian process is a collection of random variables, any finite number of which have a joint Gaussian distribution.” – <span class="citation" data-cites="rasmussen2006gaussian">[@rasmussen2006gaussian]</span></li>
<li>E.g. the function is a collection indexed by input <span class="math inline">\(x\)</span>: <span class="math display">\[  
  f(x) \sim \mathcal{GP}(m, k)
\]</span></li>
<li>It is specified by its mean function and covariance function:
<ul>
<li>Mean function <span class="math inline">\(m(x)=E[f(x)]\)</span></li>
<li>Covariance function (a.k.a. kernel) <span class="math inline">\(k(x,x')=\mathbb{E}[(f(x) - m(x))(f(x') - m(x'))]\)</span></li>
<li>A common default (e.g.&nbsp;in software) is <span class="math inline">\(m(x)=0\)</span></li>
<li>One very commonly used covariance function is the squared exponential or radial basis function (RBF)</li>
<li>We’ll see a more general form later, but for now we’re using: <span class="math inline">\(k(x, x') = \sigma^2 \exp\left(-\frac{1}{2} \|x - x'\|^2\right)\)</span></li>
<li>For now, assume data is observed without noise</li>
</ul></li>
</ul>
<p>Let’s look at a draw from the Gaussian process, i.e.&nbsp;we are just generating synthetic data from the Gaussian process model.</p>
</section>
<section id="demo-2-draw-from-a-gaussian-process" class="level2">
<h2 class="anchored" data-anchor-id="demo-2-draw-from-a-gaussian-process">Demo 2: Draw from a Gaussian process</h2>
<div id="cell-fig-demo-01-gp-draws-4" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>signal_stddev <span class="op">=</span> <span class="fl">1.0</span><span class="op">;</span> </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>kern <span class="op">=</span> signal_stddev<span class="op">**</span><span class="dv">2</span><span class="op">*</span>RBF(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  length_scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  length_scale_bounds<span class="op">=</span><span class="st">"fixed"</span>) <span class="co"># set the kernel</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>gp <span class="op">=</span> GaussianProcessRegressor(kernel<span class="op">=</span>kern)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>xobs <span class="op">=</span> np.array([<span class="fl">0.0</span>,<span class="fl">0.05</span>,<span class="fl">0.7</span>,<span class="fl">3.0</span>])<span class="op">;</span> </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>Nobs <span class="op">=</span> xobs.shape[<span class="dv">0</span>]</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>numx<span class="op">=</span><span class="dv">50</span><span class="op">;</span> </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>xstart<span class="op">=-</span><span class="fl">0.5</span><span class="op">;</span> </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>xstop<span class="op">=</span><span class="fl">3.5</span><span class="op">;</span> x<span class="op">=</span>np.linspace(</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,num<span class="op">=</span>numx)<span class="op">;</span> </span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>twosd<span class="op">=</span><span class="dv">2</span><span class="op">*</span>signal_stddev<span class="op">*</span>np.ones(numx) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>plt.fill_between(</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  x,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  twosd,</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>  <span class="op">-</span>twosd,</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"gray"</span>,</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>  alpha<span class="op">=</span><span class="fl">0.3</span>) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>xcurve<span class="op">=</span>np.linspace(start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,num<span class="op">=</span>numx)<span class="op">;</span> </span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>xall <span class="op">=</span> np.concatenate((xcurve,xobs)).reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>yall <span class="op">=</span> gp.sample_y(xall,random_state<span class="op">=</span>rng,n_samples<span class="op">=</span><span class="dv">1</span>)<span class="op">;</span> </span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>ycurve<span class="op">=</span>yall[<span class="dv">0</span>:numx]<span class="op">;</span> </span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>yobs<span class="op">=</span>yall[numx:numx<span class="op">+</span>Nobs]</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>plt.plot(xcurve,ycurve,linestyle<span class="op">=</span><span class="st">"dashed"</span>,color<span class="op">=</span><span class="st">"blue"</span>)<span class="op">;</span> </span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>plt.scatter(xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,marker<span class="op">=</span><span class="st">"x"</span>)</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> </span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>pltaxes <span class="op">=</span> plt.gca() <span class="co"># make the plot</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(<span class="st">"A draw from a Gaussian process with "</span> <span class="op">+</span> <span class="bu">str</span>(Nobs) <span class="op">+</span> <span class="st">" simulated data points"</span>) <span class="co"># make the plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-draws-4" class="quarto-float quarto-figure quarto-figure-center anchored" height="431" width="580">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-draws-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-draws-4-output-1.png" id="fig-demo-01-gp-draws-4" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-demo-01-gp-draws-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9
</figcaption>
</figure>
</div>
</div>
</div>
<div id="cell-fig-demo-01-gp-draws-5" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># just the training data</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>plt.fill_between(x,twosd,<span class="op">-</span>twosd,color<span class="op">=</span><span class="st">"gray"</span>,alpha<span class="op">=</span><span class="fl">0.3</span>) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>plt.scatter(xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,marker<span class="op">=</span><span class="st">"x"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> pltaxes <span class="op">=</span> plt.gca() <span class="co"># make the plot</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(<span class="bu">str</span>(Nobs) <span class="op">+</span> <span class="st">" simulated data points"</span>) <span class="co"># make the plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-draws-5" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-draws-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-draws-5-output-1.png" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-draws-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10: A draw from a Gaussian process with 4 simulated data points
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="note-on-dimension-of-the-input" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="note-on-dimension-of-the-input">Note on “dimension” of the input</h2>
<!--placeholder image-->

<div class="no-row-height column-margin column-container"><div id="fig-nonparametrics-dimension-data" class="quarto-float quarto-figure quarto-figure-center anchored" data-group="slides" width="53mm">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametrics-dimension-data-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-SL05.png" id="fig-nonparametrics-dimension-data" class="img-fluid figure-img" style="width:53mm" data-group="slides">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-nonparametrics-dimension-data-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11
</figcaption>
</figure>
</div><div id="fig-nonparametrics-dimension-item" class="quarto-float quarto-figure quarto-figure-center anchored" data-group="slides" width="53mm">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-nonparametrics-dimension-item-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-SL07.png" id="fig-nonparametrics-dimension-item" class="img-fluid figure-img" style="width:53mm" data-group="slides">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-nonparametrics-dimension-item-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;12
</figcaption>
</figure>
</div></div>
<ul>
<li>Let’s be careful to separate two types of “dimension”
<ul>
<li>We’re using a superscript <span class="math inline">\(y^{(1)}\)</span> to denote (M or N) number of points in the space</li>
<li>We’ll use a subscript for the (D) different elements of a point’s vector</li>
</ul></li>
<li>Note: all of our real-life examples from the start had number of inputs D &gt; 1</li>
<li>D = 1 is much easier to visualize, but might not be representative.</li>
</ul>
</section>
<section id="inference-about-unknowns-given-data." class="level2">
<h2 class="anchored" data-anchor-id="inference-about-unknowns-given-data.">Inference about Unknowns given data.</h2>
<ul>
<li><p>Let <span class="math inline">\(X\)</span> collect the N “training” data points (indexed 1 to N)</p></li>
<li><p>Let <span class="math inline">\(X'\)</span> collect the M “test” data points <span class="math inline">\(X: N\times D\)</span></p>
<ul>
<li>Where we want to evaluate the function</li>
<li>Indexed N+1:N+M</li>
</ul></li>
<li><p><span class="math inline">\(K(X,X')\)</span> os the <span class="math inline">\(N \times M\)</span> matrix with (n,m) entry <span class="math inline">\(k(x^{(n)}, x^{(N+m)})\)</span></p></li>
<li><p>Then by our model (dimension are annotated in blue) <span class="math display">\[
\begin{bmatrix}
\overbrace{f(X)}^{\color{blue}N\times 1} \\
\underbrace{f(X')}_{\color{blue}M\times 1}
\end{bmatrix}
\sim \mathcal{N}
\left(
  \underbrace{
    \begin{bmatrix}
    0_N \\
    0_M
    \end{bmatrix}}_{\color{blue}(N+M)\times 1 } ,
  \underbrace{
\begin{bmatrix}
  K(X, X) &amp; K(X, X') \\
  K(X', X) &amp; K(X', X')
\end{bmatrix}}_{
  \color{blue}(N+M)\times (N+M)
}
\right)
\]</span></p></li>
<li><p>The conditional satisfies <span class="math inline">\(f(X')\mid f(X), X, X' \sim \mathcal{N}\)</span></p>
<ul>
<li>Can compute mean (<span class="math inline">\(\color{blue}M\times 1\)</span>) &amp; covariance (<span class="math inline">\(\color{blue}M\times M\)</span>) in closed form with Gaussian facts</li>
</ul></li>
</ul>
<p><span class="math display">\[
\begin{align*}
k(x, x') = \sigma^2 \exp\left(-\tfrac{1}{2}(x - x')^2\right), \sigma^2 = 1
\end{align*}
\]</span></p>
<p>Now we can flip this around using the Bayesian machinery and use that to get information. When we don’t know the function, but we do know the data, which is real life. In real life You know the data, but you don’t know the function and you’re trying to learn that direction.</p>
<p>If we want to draw a curve we need M to be very large. And we will need to generate a <span class="math inline">\(M \times D\)</span> matrix of points.</p>
<p>So now what we’re going to do is we’re going to run Gaussian process regression on those points. And what do I mean by, say, running Gaussian process regression I’m literally talking about exactly what we just did on the slide together.</p>
<div id="cell-fig-demo-01-gp-predict" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="annotated-cell-7"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-7-1"><a href="#annotated-cell-7-1" aria-hidden="true" tabindex="-1"></a>gpfit <span class="op">=</span> GaussianProcessRegressor(</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-7" data-target-annotation="1" onclick="event.preventDefault();" href="">1</a><span id="annotated-cell-7-2" class="code-annotation-target"><a href="#annotated-cell-7-2" aria-hidden="true" tabindex="-1"></a>  kernel<span class="op">=</span>kern, optimizer<span class="op">=</span><span class="va">None</span>)</span>
<span id="annotated-cell-7-3"><a href="#annotated-cell-7-3" aria-hidden="true" tabindex="-1"></a>gpfit.fit(xobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),yobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="annotated-cell-7-4"><a href="#annotated-cell-7-4" aria-hidden="true" tabindex="-1"></a>mean_pred, stddev_pred <span class="op">=</span> gpfit.predict(</span>
<span id="annotated-cell-7-5"><a href="#annotated-cell-7-5" aria-hidden="true" tabindex="-1"></a>  xcurve.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="annotated-cell-7-6"><a href="#annotated-cell-7-6" aria-hidden="true" tabindex="-1"></a>  return_std<span class="op">=</span><span class="va">True</span>)</span>
<span id="annotated-cell-7-7"><a href="#annotated-cell-7-7" aria-hidden="true" tabindex="-1"></a>plt.scatter(</span>
<span id="annotated-cell-7-8"><a href="#annotated-cell-7-8" aria-hidden="true" tabindex="-1"></a>  xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,</span>
<span id="annotated-cell-7-9"><a href="#annotated-cell-7-9" aria-hidden="true" tabindex="-1"></a>  marker<span class="op">=</span><span class="st">"x"</span>, label<span class="op">=</span><span class="st">"training data"</span>)</span>
<span id="annotated-cell-7-10"><a href="#annotated-cell-7-10" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="annotated-cell-7-11"><a href="#annotated-cell-7-11" aria-hidden="true" tabindex="-1"></a>  xcurve,ycurve,linestyle<span class="op">=</span><span class="st">"dashed"</span>,</span>
<span id="annotated-cell-7-12"><a href="#annotated-cell-7-12" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"blue"</span>, label<span class="op">=</span><span class="st">"original function"</span>)</span>
<span id="annotated-cell-7-13"><a href="#annotated-cell-7-13" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="annotated-cell-7-14"><a href="#annotated-cell-7-14" aria-hidden="true" tabindex="-1"></a>  xcurve,mean_pred,color<span class="op">=</span><span class="st">"green"</span>, </span>
<span id="annotated-cell-7-15"><a href="#annotated-cell-7-15" aria-hidden="true" tabindex="-1"></a>  label<span class="op">=</span><span class="st">"mean predictions"</span>)</span>
<span id="annotated-cell-7-16"><a href="#annotated-cell-7-16" aria-hidden="true" tabindex="-1"></a>plt.fill_between(</span>
<span id="annotated-cell-7-17"><a href="#annotated-cell-7-17" aria-hidden="true" tabindex="-1"></a>  xcurve,mean_pred<span class="op">-</span><span class="dv">2</span><span class="op">*</span>stddev_pred,</span>
<span id="annotated-cell-7-18"><a href="#annotated-cell-7-18" aria-hidden="true" tabindex="-1"></a>  mean_pred<span class="op">+</span><span class="dv">2</span><span class="op">*</span>stddev_pred,color<span class="op">=</span><span class="st">"green"</span>,</span>
<span id="annotated-cell-7-19"><a href="#annotated-cell-7-19" aria-hidden="true" tabindex="-1"></a>  alpha<span class="op">=</span><span class="fl">0.3</span>,label<span class="op">=</span><span class="vs">r"</span><span class="op">+</span><span class="vs">/-2 std dev interval"</span>)</span>
<span id="annotated-cell-7-20"><a href="#annotated-cell-7-20" aria-hidden="true" tabindex="-1"></a>plt.legend()<span class="op">;</span> </span>
<span id="annotated-cell-7-21"><a href="#annotated-cell-7-21" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> </span>
<span id="annotated-cell-7-22"><a href="#annotated-cell-7-22" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="annotated-cell-7-23"><a href="#annotated-cell-7-23" aria-hidden="true" tabindex="-1"></a>plotout <span class="op">=</span> plt.title(<span class="st">"Gaussian process predictions"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-7" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-7" data-code-lines="2" data-code-annotation="1">NOT fitting kernel hyperparameters</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-predict" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-predict-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-predict-output-1.png" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-predict-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;13: Gaussian process predictions
</figcaption>
</figure>
</div>
</div>
</div>
<p>We’re saying hey jointly You can think of the joint distribution of the value of the value of the Gaussian process at all of these different X locations, all of these different horizontal locations.</p>
<p>What we would have is we would have everything else. In this plot we’d have The black X’s, that’s our observed data. So we totally know that.</p>
<ul>
<li>We have:
<ul>
<li><p>Our observed data, the black X’s, which is our training data.</p></li>
<li><p>Our predictions. that is the solid green line it is the mean of the Gaussian process at all of these different X locations away from the observations. We are creating it just like we created our draws or like we created our, our big prior Gaussian process before.</p></li>
<li><p>We’re saying, hey, what’s the mean? In this case, it’s non-trivial. It’s not just zero because it’s conditioned on the data.</p></li>
<li><p>We have our uncertainties, our plus or minus two standard deviation interval.</p></li>
</ul></li>
</ul>
<p>Now, before, this was pretty boring. When we were first generating things a priori when we didn’t have any data. The plus or minus two standard deviation interval is just plus or minus two everywhere. Now, because we’ve observed some data, it’s not trivial. We’re actually seeing that it differs depending on where we are in the space.</p>
<p>And I just want to point out some things that might be considered useful about this.</p>
<ol type="1">
<li>When I’m really near my data points
<ul>
<li>we see there’s very little uncertainty.</li>
<li>very little uncertainty going between them. Because this is basically the only way we can smoothly and quickly go between these two points.</li>
</ul></li>
<li>when we get much farther away we have a lot of uncertainty because there’s a good number of different ways that we could have gotten between these two points. It makes sense that our uncertainty would be larger As we get farther away.</li>
</ol>
<div id="cell-fig-demo-01-gp-predict-2" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="annotated-cell-8"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><a class="code-annotation-anchor" data-target-cell="annotated-cell-8" data-target-annotation="1" onclick="event.preventDefault();" href="">1</a><span id="annotated-cell-8-1" class="code-annotation-target"><a href="#annotated-cell-8-1" aria-hidden="true" tabindex="-1"></a>gpfit <span class="op">=</span> GaussianProcessRegressor(kernel<span class="op">=</span>kern, optimizer<span class="op">=</span><span class="va">None</span>)</span>
<span id="annotated-cell-8-2"><a href="#annotated-cell-8-2" aria-hidden="true" tabindex="-1"></a>gpfit.fit(xobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),yobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="annotated-cell-8-3"><a href="#annotated-cell-8-3" aria-hidden="true" tabindex="-1"></a>mean_pred, stddev_pred <span class="op">=</span> gpfit.predict(xcurve.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), return_std<span class="op">=</span><span class="va">True</span>)</span>
<span id="annotated-cell-8-4"><a href="#annotated-cell-8-4" aria-hidden="true" tabindex="-1"></a>y_all_draws<span class="op">=</span>gpfit.sample_y(xcurve.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),random_state<span class="op">=</span>rng,n_samples<span class="op">=</span><span class="dv">3</span>)</span>
<span id="annotated-cell-8-5"><a href="#annotated-cell-8-5" aria-hidden="true" tabindex="-1"></a>color_options<span class="op">=</span>[<span class="st">"red"</span>,<span class="st">"blue"</span>,<span class="st">"orange"</span>]</span>
<span id="annotated-cell-8-6"><a href="#annotated-cell-8-6" aria-hidden="true" tabindex="-1"></a>plt.scatter(xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,marker<span class="op">=</span><span class="st">"x"</span>, label<span class="op">=</span><span class="st">"training data"</span>)</span>
<span id="annotated-cell-8-7"><a href="#annotated-cell-8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="annotated-cell-8-8"><a href="#annotated-cell-8-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, ydraw <span class="kw">in</span> <span class="bu">enumerate</span>(y_all_draws.T): </span>
<span id="annotated-cell-8-9"><a href="#annotated-cell-8-9" aria-hidden="true" tabindex="-1"></a>  plt.plot(xcurve,ydraw,label<span class="op">=</span><span class="st">"random draw"</span>,color<span class="op">=</span>color_options[np.mod(i,<span class="dv">3</span>)])</span>
<span id="annotated-cell-8-10"><a href="#annotated-cell-8-10" aria-hidden="true" tabindex="-1"></a>plt.fill_between(</span>
<span id="annotated-cell-8-11"><a href="#annotated-cell-8-11" aria-hidden="true" tabindex="-1"></a>  xcurve,mean_pred<span class="op">-</span><span class="dv">2</span><span class="op">*</span>stddev_pred,mean_pred<span class="op">+</span><span class="dv">2</span><span class="op">*</span>stddev_pred,</span>
<span id="annotated-cell-8-12"><a href="#annotated-cell-8-12" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"green"</span>,alpha<span class="op">=</span><span class="fl">0.3</span>,label<span class="op">=</span><span class="vs">r"</span><span class="op">+</span><span class="vs">/-2 std dev interval"</span>)</span>
<span id="annotated-cell-8-13"><a href="#annotated-cell-8-13" aria-hidden="true" tabindex="-1"></a>plt.legend()<span class="op">;</span> </span>
<span id="annotated-cell-8-14"><a href="#annotated-cell-8-14" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> </span>
<span id="annotated-cell-8-15"><a href="#annotated-cell-8-15" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="annotated-cell-8-16"><a href="#annotated-cell-8-16" aria-hidden="true" tabindex="-1"></a>plotout <span class="op">=</span> plt.title(<span class="st">"Random draws of f from the predictive distribution"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-8" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-8" data-code-lines="1" data-code-annotation="1">NOT fitting kernel hyperparameters</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-predict-2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-predict-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-predict-2-output-1.png" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-predict-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;14: Gaussian process predictions with noise
</figcaption>
</figure>
</div>
</div>
</div>
<p>Now we have these four data points. And so we’ve constrained our random draws of the functions to go through the four data points.</p>
<p>And here are just three draws from it. Because again, it’s just a multivariate Gaussian is all that we’re doing here. We’re just making this multivariate Gaussian the same way we did before by having draws at many dense x values at many dense input values. And so we can do the same thing here. It’s just a different multivariate Gaussian.</p>
<p>This is meant to express Our uncertainty and our best guess, our best guess could be seen as this mean This posterior mean. So these are the mean predictions this solid green line. So that could be thought of as our best guess. And then our uncertainty around that</p>
<p>Is the plus or minus two standard deviation interval conditional on the data that we’ve seen. And now we’re going to dig a little bit more. Into some choices we’ve made and what are their implications.</p>
</section>
<section id="sec-squared-exponential-kernel-revisited" class="level2">
<h2 class="anchored" data-anchor-id="sec-squared-exponential-kernel-revisited">Squared exponential kernel revisited</h2>
<p>What if we happened to measure our data on a different scale?</p>
<p>We’ve been using this particular kernel:</p>
<p><span class="math display">\[
\begin{align*}
k(x, x') = \sigma^2 \exp\left(-\tfrac{1}{2}(x - x')^2\right), \sigma^2 = 1
\end{align*}
\]</span></p>
<ul>
<li>What do we expect from the scale of f(x) a priori?
<ul>
<li>At one x, with ~95% probability a priori,f(x) is in the interval <span class="math inline">\([-2\sigma,2\sigma]\)</span>.</li>
<li>Marginal variance cannot increase with data</li>
</ul></li>
<li>What counts as “close” in x?</li>
</ul>
<p><span class="math display">\[
\exp(-\tfrac{1}{2}(2)^2) \approx 0.14 \quad \exp(-\tfrac{1}{3}(3)^2) \approx 0.11 \quad  \exp(-\tfrac{1}{2}(4)^2) \approx 0.0034 \qquad
\]</span></p>
<div id="cell-fig-demo-01-gp-predict-3" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="annotated-cell-9"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-9-1"><a href="#annotated-cell-9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># observed y values are 100 times what they were in the nice example</span></span>
<span id="annotated-cell-9-2"><a href="#annotated-cell-9-2" aria-hidden="true" tabindex="-1"></a>gpfit <span class="op">=</span> GaussianProcessRegressor(</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-9" data-target-annotation="1" onclick="event.preventDefault();" href="">1</a><span id="annotated-cell-9-3" class="code-annotation-target"><a href="#annotated-cell-9-3" aria-hidden="true" tabindex="-1"></a>  kernel<span class="op">=</span>kern, optimizer<span class="op">=</span><span class="va">None</span>)</span>
<span id="annotated-cell-9-4"><a href="#annotated-cell-9-4" aria-hidden="true" tabindex="-1"></a>gpfit.fit(xobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),</span>
<span id="annotated-cell-9-5"><a href="#annotated-cell-9-5" aria-hidden="true" tabindex="-1"></a>  <span class="dv">100</span><span class="op">*</span>yobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="annotated-cell-9-6"><a href="#annotated-cell-9-6" aria-hidden="true" tabindex="-1"></a>mean_pred, stddev_pred <span class="op">=</span> gpfit.predict(</span>
<span id="annotated-cell-9-7"><a href="#annotated-cell-9-7" aria-hidden="true" tabindex="-1"></a>  xcurve.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="annotated-cell-9-8"><a href="#annotated-cell-9-8" aria-hidden="true" tabindex="-1"></a>  return_std<span class="op">=</span><span class="va">True</span>)</span>
<span id="annotated-cell-9-9"><a href="#annotated-cell-9-9" aria-hidden="true" tabindex="-1"></a>plt.scatter(xobs,<span class="dv">100</span><span class="op">*</span>yobs,color<span class="op">=</span><span class="st">"black"</span>,</span>
<span id="annotated-cell-9-10"><a href="#annotated-cell-9-10" aria-hidden="true" tabindex="-1"></a>marker<span class="op">=</span><span class="st">"x"</span>, label<span class="op">=</span><span class="st">"training data"</span>)</span>
<span id="annotated-cell-9-11"><a href="#annotated-cell-9-11" aria-hidden="true" tabindex="-1"></a>plt.plot(xcurve,<span class="dv">100</span><span class="op">*</span>ycurve,linestyle<span class="op">=</span><span class="st">"dashed"</span>,</span>
<span id="annotated-cell-9-12"><a href="#annotated-cell-9-12" aria-hidden="true" tabindex="-1"></a>color<span class="op">=</span><span class="st">"blue"</span>, label<span class="op">=</span><span class="st">"original function"</span>)</span>
<span id="annotated-cell-9-13"><a href="#annotated-cell-9-13" aria-hidden="true" tabindex="-1"></a>plt.plot(xcurve,mean_pred,color<span class="op">=</span><span class="st">"green"</span>, </span>
<span id="annotated-cell-9-14"><a href="#annotated-cell-9-14" aria-hidden="true" tabindex="-1"></a>label<span class="op">=</span><span class="st">"mean predictions"</span>)</span>
<span id="annotated-cell-9-15"><a href="#annotated-cell-9-15" aria-hidden="true" tabindex="-1"></a>plt.fill_between(</span>
<span id="annotated-cell-9-16"><a href="#annotated-cell-9-16" aria-hidden="true" tabindex="-1"></a>  xcurve,</span>
<span id="annotated-cell-9-17"><a href="#annotated-cell-9-17" aria-hidden="true" tabindex="-1"></a>  mean_pred<span class="op">-</span><span class="dv">2</span><span class="op">*</span>stddev_pred,mean_pred<span class="op">+</span><span class="dv">2</span><span class="op">*</span>stddev_pred,</span>
<span id="annotated-cell-9-18"><a href="#annotated-cell-9-18" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"green"</span>,alpha<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="annotated-cell-9-19"><a href="#annotated-cell-9-19" aria-hidden="true" tabindex="-1"></a>  label<span class="op">=</span><span class="vs">r"</span><span class="op">+</span><span class="vs">/-2 std dev interval"</span>)</span>
<span id="annotated-cell-9-20"><a href="#annotated-cell-9-20" aria-hidden="true" tabindex="-1"></a>plt.legend()<span class="op">;</span> </span>
<span id="annotated-cell-9-21"><a href="#annotated-cell-9-21" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> </span>
<span id="annotated-cell-9-22"><a href="#annotated-cell-9-22" aria-hidden="true" tabindex="-1"></a>plt.ylim(</span>
<span id="annotated-cell-9-23"><a href="#annotated-cell-9-23" aria-hidden="true" tabindex="-1"></a>  (<span class="op">-</span><span class="dv">100</span><span class="op">*</span><span class="dv">3</span><span class="op">*</span>signal_stddev, <span class="dv">100</span><span class="op">*</span><span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="annotated-cell-9-24"><a href="#annotated-cell-9-24" aria-hidden="true" tabindex="-1"></a>plotout <span class="op">=</span> plt.title(<span class="st">"Gaussian process predictions"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-9" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-9" data-code-lines="3" data-code-annotation="1">NOT fitting kernel hyperparameters</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-predict-3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-predict-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-predict-3-output-1.png" width="596" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-predict-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15: issues when we scale the y values by 100
</figcaption>
</figure>
</div>
</div>
</div>
<p>In this sample the <span class="math inline">\(+/-2 \sigma\)</span> confidence interval is now 100 times smaller than in the nice example, and the mean prediction is also 100 times smaller than in the nice example. So while it is still there it seems to have vanished.</p>
<div id="cell-fig-demo-01-gp-predict-4" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># observed x values are 100 times what they were in the nice example</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>gpfit <span class="op">=</span> GaussianProcessRegressor(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  kernel<span class="op">=</span>kern, optimizer<span class="op">=</span><span class="va">None</span>) <span class="co"># NOT fitting kernel hyperparameters</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>gpfit.fit(<span class="dv">100</span><span class="op">*</span>xobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),yobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>xall<span class="op">=</span>np.sort(np.concatenate((xcurve,xobs)))</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>mean_pred, stddev_pred <span class="op">=</span> gpfit.predict(</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="dv">100</span><span class="op">*</span>xall.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  return_std<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>plt.scatter(<span class="dv">100</span><span class="op">*</span>xobs,</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  yobs,color<span class="op">=</span><span class="st">"black"</span>,marker<span class="op">=</span><span class="st">"x"</span>, </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  label<span class="op">=</span><span class="st">"training data"</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  <span class="dv">100</span><span class="op">*</span>xall,mean_pred,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"green"</span>, label<span class="op">=</span><span class="st">"mean predictions"</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>  <span class="dv">100</span><span class="op">*</span>xcurve,ycurve,</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  linestyle<span class="op">=</span><span class="st">"dashed"</span>,</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"blue"</span>, </span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>  label<span class="op">=</span><span class="st">"original function"</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>plt.fill_between(</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>  <span class="dv">100</span><span class="op">*</span>xall,mean_pred<span class="op">-</span><span class="dv">2</span><span class="op">*</span>stddev_pred,</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  mean_pred<span class="op">+</span><span class="dv">2</span><span class="op">*</span>stddev_pred,color<span class="op">=</span><span class="st">"green"</span>,</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>  alpha<span class="op">=</span><span class="fl">0.3</span>,label<span class="op">=</span><span class="vs">r"</span><span class="op">+</span><span class="vs">/-2 std dev interval"</span>)</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>plt.legend()<span class="op">;</span> </span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>plt.xlim([<span class="dv">100</span><span class="op">*</span>xstart,<span class="dv">100</span><span class="op">*</span>xstop])<span class="op">;</span> </span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>plotout <span class="op">=</span> plt.title(<span class="st">"Gaussian process predictions"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-demo-01-gp-predict-4" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-01-gp-predict-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-01-gp-predict-4-output-1.png" width="582" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-01-gp-predict-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;16: Issue when we scale the x values by 100
</figcaption>
</figure>
</div>
</div>
</div>
<p>in this example the points are now so far apart that at most point on the graph we are quickly jumping to the prior and getting the maximum uncertainty.</p>
<div id="lst-estimation" class="cell listing quarto-float quarto-figure quarto-figure-left anchored" data-execution_count="11">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst quarto-uncaptioned" id="lst-estimation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2
</figcaption>
<div aria-describedby="lst-estimation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(np.exp(<span class="op">-</span><span class="fl">0.5</span><span class="op">*</span><span class="fl">2.0</span><span class="op">**</span><span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0.13533528323661267</code></pre>
</div>
</div>
</figure>
</div>
<ul>
<li>What can we do to handle different x and f(x) scales?
<ul>
<li>Normalization in y can help; in x, can still be hiccups</li>
</ul></li>
<li>A common option in practice and in software is to fit the hyperparameters of a more general squared exponential kernel from data
<ul>
<li>More general form of the squared exponential:</li>
</ul></li>
</ul>
<p><span id="eq-squared-exponential-kernel-general"><span class="math display">\[
k(x, x') = \underbrace{\sigma^2}_{\color{red}\text{signal variance}} \exp\left(-\frac{1}{2}\sum_{d=1}^D \frac{(x_d - x'_d)^2}{\underbrace{l_d^2}_{\color{red}\text{length scale}}}\right), \sigma^2 = 1
\tag{1}\]</span></span></p>
<ul>
<li><p>Parameters (here, f) parametrize the distribution of the data. If we knew them, we could generate the data.</p>
<ul>
<li>GPs: <em>nonparametric</em> model: infinite # of latent params</li>
</ul></li>
<li><p><em>Hyperparameters</em> parametrize the distribution of the parameters. If known, we could generate the parameters.</p></li>
<li><p>Algorithm:</p>
<ul>
<li>Fit a value for the hyperparameters using the data.</li>
<li>Given those values, now compute and report the mean and uncertainty intervals</li>
</ul></li>
</ul>
<section id="demo-3-learning-the-signal-and-the-length-scale-and" class="level3">
<h3 class="anchored" data-anchor-id="demo-3-learning-the-signal-and-the-length-scale-and">Demo 3 – Learning the signal and the length scale and</h3>
<div id="cell-fig-demo-03-gp-data-and-ground-truth" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>signal_stddev <span class="op">=</span> <span class="fl">2.0</span><span class="op">;</span> </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>kern <span class="op">=</span> signal_stddev<span class="op">**</span><span class="dv">2</span><span class="op">*</span>RBF(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  length_scale<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  length_scale_bounds<span class="op">=</span><span class="st">"fixed"</span>) <span class="co"># set the kernel</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>gp <span class="op">=</span> GaussianProcessRegressor(kernel<span class="op">=</span>kern)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>xstart<span class="op">=-</span><span class="fl">0.5</span><span class="op">;</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>xstop<span class="op">=</span><span class="fl">3.5</span><span class="op">;</span> </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>Nobs <span class="op">=</span> <span class="dv">10</span><span class="op">;</span> </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>xobs <span class="op">=</span> rng.uniform(xstart,xstop,size<span class="op">=</span>Nobs)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>numx<span class="op">=</span><span class="dv">200</span><span class="op">;</span> </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>x<span class="op">=</span>np.linspace(start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,num<span class="op">=</span>numx)<span class="op">;</span> </span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>twosd<span class="op">=</span><span class="dv">2</span><span class="op">*</span>signal_stddev<span class="op">*</span>np.ones(numx) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>plt.fill_between(x,twosd,<span class="op">-</span>twosd,color<span class="op">=</span><span class="st">"gray"</span>,alpha<span class="op">=</span><span class="fl">0.3</span>) <span class="co"># coloring 2 std dev range</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>xcurve<span class="op">=</span>np.linspace(start<span class="op">=</span>xstart,stop<span class="op">=</span>xstop,num<span class="op">=</span>numx)<span class="op">;</span> </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>xall <span class="op">=</span> np.concatenate((xcurve,xobs)).reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>yall <span class="op">=</span> gp.sample_y(xall,random_state<span class="op">=</span>rng,n_samples<span class="op">=</span><span class="dv">1</span>)<span class="op">;</span> </span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>ycurve<span class="op">=</span>yall[<span class="dv">0</span>:numx]<span class="op">;</span> </span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>yobs<span class="op">=</span>yall[numx:numx<span class="op">+</span>Nobs]</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>plt.plot(xcurve,ycurve,linestyle<span class="op">=</span><span class="st">"dashed"</span>,color<span class="op">=</span><span class="st">"blue"</span>)<span class="op">;</span> </span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>plt.scatter(xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,marker<span class="op">=</span><span class="st">"x"</span>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> </span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>pltaxes <span class="op">=</span> plt.gca() <span class="co"># make the plot</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>  <span class="st">"A draw from a Gaussian process with "</span> <span class="op">+</span> <span class="bu">str</span>(Nobs) <span class="op">+</span> </span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>  <span class="st">" simulated data points"</span>) <span class="co"># make the plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-demo-03-gp-data-and-ground-truth" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo-03-gp-data-and-ground-truth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo-03-gp-data-and-ground-truth-output-1.png" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo-03-gp-data-and-ground-truth-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17: A draw from a Gaussian process with 10 simulated data points
</figcaption>
</figure>
</div>
</div>
</div>
<p>we have drawn 10 points and the ground truth</p>
<p>but what goes to the algorithm is just the data points as follows:</p>
<div id="cell-fig-demo3-gp-data" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># just the training data</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>plt.scatter(xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,marker<span class="op">=</span><span class="st">"x"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> pltaxes <span class="op">=</span> plt.gca() <span class="co"># make the plot</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> plt.title(<span class="bu">str</span>(Nobs) <span class="op">+</span> <span class="st">" simulated data points"</span>) <span class="co"># make the plot</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div id="fig-demo3-gp-data" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo3-gp-data-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo3-gp-data-output-1.png" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo3-gp-data-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;18: the data
</figcaption>
</figure>
</div>
</div>
</div>
<div id="cell-fig-demo3-gp-algorithm" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="annotated-cell-14"><pre class="sourceCode python code-annotation-code code-with-copy code-annotated"><code class="sourceCode python"><span id="annotated-cell-14-1"><a href="#annotated-cell-14-1" aria-hidden="true" tabindex="-1"></a>signal_stddev_init <span class="op">=</span> <span class="fl">1.0</span><span class="op">;</span> </span>
<span id="annotated-cell-14-2"><a href="#annotated-cell-14-2" aria-hidden="true" tabindex="-1"></a>kern_fit <span class="op">=</span> signal_stddev_init<span class="op">**</span><span class="dv">2</span><span class="op">*</span>RBF(</span>
<span id="annotated-cell-14-3"><a href="#annotated-cell-14-3" aria-hidden="true" tabindex="-1"></a>  length_scale<span class="op">=</span><span class="fl">1.0</span>,length_scale_bounds<span class="op">=</span>(<span class="fl">0.01</span>, <span class="dv">100</span>))</span>
<span id="annotated-cell-14-4"><a href="#annotated-cell-14-4" aria-hidden="true" tabindex="-1"></a>gpfit <span class="op">=</span> GaussianProcessRegressor(</span>
<a class="code-annotation-anchor" data-target-cell="annotated-cell-14" data-target-annotation="1" onclick="event.preventDefault();" href="">1</a><span id="annotated-cell-14-5" class="code-annotation-target"><a href="#annotated-cell-14-5" aria-hidden="true" tabindex="-1"></a>  kernel<span class="op">=</span>kern_fit,n_restarts_optimizer<span class="op">=</span><span class="dv">10</span>)</span>
<span id="annotated-cell-14-6"><a href="#annotated-cell-14-6" aria-hidden="true" tabindex="-1"></a>gpfit.fit(xobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>),yobs.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>))<span class="op">;</span></span>
<span id="annotated-cell-14-7"><a href="#annotated-cell-14-7" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Original kernel: "</span> <span class="op">+</span> <span class="bu">str</span>(kern))<span class="op">;</span> </span>
<span id="annotated-cell-14-8"><a href="#annotated-cell-14-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Fit kernel: "</span> <span class="op">+</span> <span class="bu">str</span>(gpfit.kernel_))</span>
<span id="annotated-cell-14-9"><a href="#annotated-cell-14-9" aria-hidden="true" tabindex="-1"></a>mean_pred, stddev_pred <span class="op">=</span> gpfit.predict(</span>
<span id="annotated-cell-14-10"><a href="#annotated-cell-14-10" aria-hidden="true" tabindex="-1"></a>  xcurve.reshape(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>), return_std<span class="op">=</span><span class="va">True</span>)</span>
<span id="annotated-cell-14-11"><a href="#annotated-cell-14-11" aria-hidden="true" tabindex="-1"></a>plt.scatter(</span>
<span id="annotated-cell-14-12"><a href="#annotated-cell-14-12" aria-hidden="true" tabindex="-1"></a>  xobs,yobs,color<span class="op">=</span><span class="st">"black"</span>,</span>
<span id="annotated-cell-14-13"><a href="#annotated-cell-14-13" aria-hidden="true" tabindex="-1"></a>  marker<span class="op">=</span><span class="st">"x"</span>, label<span class="op">=</span><span class="st">"training data"</span>)</span>
<span id="annotated-cell-14-14"><a href="#annotated-cell-14-14" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="annotated-cell-14-15"><a href="#annotated-cell-14-15" aria-hidden="true" tabindex="-1"></a>  xcurve,ycurve,linestyle<span class="op">=</span><span class="st">"dashed"</span>,</span>
<span id="annotated-cell-14-16"><a href="#annotated-cell-14-16" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"blue"</span>, label<span class="op">=</span><span class="st">"original function"</span>)</span>
<span id="annotated-cell-14-17"><a href="#annotated-cell-14-17" aria-hidden="true" tabindex="-1"></a>plt.plot(</span>
<span id="annotated-cell-14-18"><a href="#annotated-cell-14-18" aria-hidden="true" tabindex="-1"></a>  xcurve,mean_pred,color<span class="op">=</span><span class="st">"green"</span>, </span>
<span id="annotated-cell-14-19"><a href="#annotated-cell-14-19" aria-hidden="true" tabindex="-1"></a>  label<span class="op">=</span><span class="st">"mean predictions"</span>)</span>
<span id="annotated-cell-14-20"><a href="#annotated-cell-14-20" aria-hidden="true" tabindex="-1"></a>plt.fill_between(</span>
<span id="annotated-cell-14-21"><a href="#annotated-cell-14-21" aria-hidden="true" tabindex="-1"></a>  xcurve,mean_pred<span class="op">-</span><span class="dv">2</span><span class="op">*</span>stddev_pred,</span>
<span id="annotated-cell-14-22"><a href="#annotated-cell-14-22" aria-hidden="true" tabindex="-1"></a>  mean_pred<span class="op">+</span><span class="dv">2</span><span class="op">*</span>stddev_pred,</span>
<span id="annotated-cell-14-23"><a href="#annotated-cell-14-23" aria-hidden="true" tabindex="-1"></a>  color<span class="op">=</span><span class="st">"green"</span>,alpha<span class="op">=</span><span class="fl">0.3</span>,</span>
<span id="annotated-cell-14-24"><a href="#annotated-cell-14-24" aria-hidden="true" tabindex="-1"></a>  label<span class="op">=</span><span class="vs">r"</span><span class="op">+</span><span class="vs">/-2 std dev interval"</span>)</span>
<span id="annotated-cell-14-25"><a href="#annotated-cell-14-25" aria-hidden="true" tabindex="-1"></a>plt.legend()<span class="op">;</span> </span>
<span id="annotated-cell-14-26"><a href="#annotated-cell-14-26" aria-hidden="true" tabindex="-1"></a>plt.xlim([xstart,xstop])<span class="op">;</span> </span>
<span id="annotated-cell-14-27"><a href="#annotated-cell-14-27" aria-hidden="true" tabindex="-1"></a>plt.ylim((<span class="op">-</span><span class="dv">3</span><span class="op">*</span>signal_stddev,<span class="dv">3</span><span class="op">*</span>signal_stddev))<span class="op">;</span> </span>
<span id="annotated-cell-14-28"><a href="#annotated-cell-14-28" aria-hidden="true" tabindex="-1"></a>plotout <span class="op">=</span> plt.title(<span class="st">"Gaussian process predictions"</span>)</span><div class="code-annotation-gutter-bg"></div><div class="code-annotation-gutter"></div></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-annotation">
<dl class="code-annotation-container-grid">
<dt data-target-cell="annotated-cell-14" data-target-annotation="1">1</dt>
<dd>
<span data-code-cell="annotated-cell-14" data-code-lines="5" data-code-annotation="1">now we ARE fitting the hyperparameters</span>
</dd>
</dl>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Original kernel: 2**2 * RBF(length_scale=0.3)
Fit kernel: 1.68**2 * RBF(length_scale=0.24)</code></pre>
</div>
<div class="cell-output cell-output-display">
<div id="fig-demo3-gp-algorithm" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-demo3-gp-algorithm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-L01_files/figure-html/fig-demo3-gp-algorithm-output-2.png" width="580" height="431" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-demo3-gp-algorithm-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;19: GP scaled by the algorithm
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="what-next" class="level2">
<h2 class="anchored" data-anchor-id="what-next">What next</h2>
<ul>
<li>adding observation noise variance</li>
<li>extrapolation</li>
<li>high-dimensional inputs</li>
</ul>
<p>Always ask: What uncertainty are we quantifying? <a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a></p>
</section>
<section id="observation-noise" class="level2">
<h2 class="anchored" data-anchor-id="observation-noise">Observation noise</h2>
<ul>
<li>So far we’ve been assuming that we observed f(x) directly</li>
<li>But often the actual observation y has additional noise:</li>
</ul>
<p><span class="math display">\[
f \sim \mathcal{GP}(m, k) \quad y^{(n)} = f(x) + \epsilon^{(n)}, \qquad \epsilon^{(n)} \stackrel{iid}{\sim} \mathcal{N}(0, \tau^2)
\]</span></p>
<ul>
<li>we observe <span class="math inline">\(\{(x^{(n)}, y^{(n)})\}_{n=1}^N\)</span> and want to learn the latent <span class="math inline">\(f\)</span>
<ul>
<li>The <span class="math inline">\(y\)</span> are multivariate-Gaussian distributed with mean and covariance summed.</li>
<li>The mean of <span class="math inline">\(y^{(n)}\)</span> is given by the function <span class="math inline">\(m(x^{(n)})\)</span>. and <span class="math inline">\(Cov(y^{(n)},y^{(n')})=k(x^{(n)},x^{(n')})+\tau^2\mathbb{1}_{n=n'}\)</span></li>
<li>The covariance is given by the kernel function:</li>
</ul></li>
</ul>
<p><span class="math display">\[
y \sim \mathcal{N}(f(x), \tau^2 I)
\]</span></p>
<p>Before: <span class="math display">\[
\begin{bmatrix}
f(X) \\
f(X')
\end{bmatrix}
\sim \mathcal{N}
  \left(    
    \begin{bmatrix}
      0_N \\
      0_M
      \end{bmatrix}
    \begin{bmatrix}
    K(X, X) &amp; K(X, X') \\
    K(X', X) &amp; K(X', X')
    \end{bmatrix}    
  \right)
\]</span> Now :</p>
<p><span class="math display">\[
\begin{bmatrix}
y^{(1:N)} \\
f(X')
\end{bmatrix}
\sim \mathcal{N}
  \left(    
    \begin{bmatrix}
      0_N \\
      0_M
      \end{bmatrix}
    \begin{bmatrix}
    K(X, X)+ \tau \mathbb{I} &amp; K(X, X') \\
    K(X', X) &amp; K(X', X')
    \end{bmatrix}    
  \right)
\]</span></p>
</section>
<section id="what-uncertainty-are-we-quantifying" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="what-uncertainty-are-we-quantifying">What uncertainty are we quantifying?</h2>
<ul>
<li>It’s worth being aware that data science (ML/stats/AI) often overloads common colloquial terms with terms of art
<ul>
<li>E.g. “significance”, “bias”, “generalization”</li>
<li>Every precise use of “uncertainty” has this issue
<ul>
<li>E.g. frequentist sampling, Bayesian, etc.</li>
</ul></li>
<li>We should always make sure we can distinguish what is, and what is not, covered by the term of art</li>
</ul></li>
<li>A standard setup (our setup so far):
<ul>
<li>We model the data as generated according to a GP with squared exponential kernel and observation noise</li>
<li>We fit the hyperparameters (the signal variance, the length scale(s), and the noise variance) to single values</li>
<li>The reported uncertainties are what result when the GP model and fitted hyperparameters are exactly correct</li>
</ul>
</li>
</ul>
<div class="no-row-height column-margin column-container"><span class="">Are there other uncertainties that aren’t being quantified here?</span></div></section>
<section id="sec-other-sources-of-uncertainty" class="level2">
<h2 class="anchored" data-anchor-id="sec-other-sources-of-uncertainty">Some other sources of uncertainty</h2>
<ul>
<li>There may be multiple sets of substantively different hyperparameter values that are both plausible and consistent with the observed data</li>
<li>What can we do? First: unit test, plot, sense check!
<ul>
<li>Ask what is possible to learn with the data available</li>
<li>Multiple random restarts: plot the results</li>
<li>Bayesian model of the hyperparameters</li>
</ul></li>
<li>A GP with your mean &amp; kernel may be meaningfully misspecified for the data
<ul>
<li>Box: “All models are wrong, but some are useful”</li>
<li>What can we do? First: unit test, plot, sense check!
<ul>
<li>Can change the mean and/or kernel
<ul>
<li>E.g. local/heteroskedastic models, periodic kernels, linear mean function, many many more</li>
</ul></li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="sec-extrapolation" class="level2">
<h2 class="anchored" data-anchor-id="sec-extrapolation">Extrapolation</h2>
<p>Extrapolation: Estimation/prediction beyond the observed data • Compare to interpolation: estimation/prediction within the observed data • When using GPs with a squared exponential kernel: • Data points that are more than a handful of length scales from other data points will revert to prior behavior • Note: extrapolation isn’t a special issue unique to GPs. It’s a fundamentally hard problem for all data analysis methods • To extrapolate, you need to make assumptions • When you have domain knowledge of a system, you might be able to use it to extrapolate • When you’re letting a machine learning method use its defaults, it’s making assumptions. Do you know what those assumptions are?</p>
</section>
<section id="more-than-one-input" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="more-than-one-input">More than one input</h2>

<div class="no-row-height column-margin column-container"><div id="fig-multiple-inputs" class="quarto-float quarto-figure quarto-figure-center anchored" data-group="slides">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-multiple-inputs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="C6-SL09.png" class="img-fluid figure-img" style="width:53mm" data-group="slides">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-multiple-inputs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;20: histogram of squared inter-point distances
</figcaption>
</figure>
</div></div><ul>
<li>Our illustrations have almost all been for one input so far</li>
<li>But in real life, it’s typical to have more than one input</li>
<li>What could go wrong? Previous lessons apply, but also:
<ul>
<li>Possibly different length scales. Check defaults.</li>
<li>Regression in high dimensions is a fundamentally hard problem (without additional assumptions)</li>
</ul></li>
<li>All points are “far away” in high dimensions. Illustration:</li>
<li>Uniformly randomly sample 10,000 points on <span class="math inline">\([0,1]^D\)</span></li>
<li>Make a histogram of squared inter-point distances</li>
<li>Recall: points “far” from data default to the prior mean and variance.</li>
</ul>
</section>
<section id="sec-high-points" class="level2">
<h2 class="anchored" data-anchor-id="sec-high-points">Some high points of what got cut for time</h2>
<ul>
<li>We ran out of time! Here are some high-level summary points beyond what we discussed together:
<ul>
<li>There are other challenges with many inputs, both conceptual and practical</li>
<li>Running time for GP regression can be an issue with a large number of training data points
<ul>
<li>In particular, the matrix inverse can be expensive</li>
<li>There are incredibly many papers about fast approximations to the exact Gaussian process
<ul>
<li>Each approximation has pros and cons</li>
</ul></li>
</ul></li>
</ul></li>
<li>Bayesian optimization inherits many of the pros and cons of Gaussian processes for regression
<ul>
<li>Exercise: once you learn about Bayesian optimization, think about how the pros and cons we discussed together might translate there</li>
</ul></li>
</ul>
</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>This is key as in real life there are many sources of uncertainty and in our models we only get to quantify one!<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>